### PR TITLE
Expose `IO#timeout` and `IO#timeout=` if available

### PR DESCRIPTION
**Pull request by casperisfine**
Ref: https://bugs.ruby-lang.org/issues/18630

FYI: @ioquatix

### CODE REVIEW

**Comment by casperisfine** [comment]
Actually, might not be that simple, as the buffering layer breaks on a timeout:

```
RedisClient::SSLConnectionTest#test_tcp_downstream_latency:
TypeError: no implicit conversion of Integer into Hash
    /opt/rubies/3.2.2/lib/ruby/3.2.0/openssl/buffering.rb:247:in `>='
    /opt/rubies/3.2.2/lib/ruby/3.2.0/openssl/buffering.rb:247:in `gets'
```

I'll have to dig deeper.

**Comment by casperisfine** [comment]
Yeah, definitely doesn't work, `OpenSSL` buffering uses `ossl_ssl_read_internal`, which isn't aware of `IO#timeout` it seems.

**Comment by ioquatix** [comment]
https://github.com/ruby/openssl/blob/fac5cac3a80166951291e351f0015c0df87ac771/ext/openssl/ossl_ssl.c#L1744

Maybe we need to set it to nonblock mode? In theory, it should use the default timeout.

**Comment by casperisfine** [comment]
> Maybe we need to set it to nonblock mode?

Perhaps? I'm not quite familiar enough with that code.

**Comment by ioquatix** [comment]
@rhenium is there a chance you can help look into this? I'm happy to help.

**Comment by rhenium** [comment]
There seem to be two issues before the buffering layer.

IO::Timeout isn't raised by OpenSSL::SSL at all. `rb_io_maybe_wait_readable()` and `rb_io_maybe_wait_writable()` on timeout return 0. This return value is currently discarded.

Also, a single blocking operation on TLS, such as `#connect` and `#sysread`, can involve multiple blocking operations on the underlying socket. This would mean we can't rely on RUBY_IO_TIMEOUT_DEFAULT, but have to get `rb_io_timeout()` and subtract elapsed time before waiting again.

**Comment by rhenium** [comment]
> https://github.com/ruby/openssl/blob/fac5cac3a80166951291e351f0015c0df87ac771/ext/openssl/ossl_ssl.c#L1744
> 
> Maybe we need to set it to nonblock mode? In theory, it should use the default timeout.

The underlying socket is set to non-blocking mode by SSLSocket#initialize. We needed this for implementing SSLSocket#read_nonblock and also because we didn't want blocking IO within libssl while holding the GVL.

File: lib/openssl/ssl.rb [hunk 1]
```diff
@@ -421,6 +421,18 @@ def session
...
+          @io.timeout
+        end
+      end
+
+      if IO.method_defined?(:timeout=)
+        def timeout=(timeout)
+          @io.timeout = timeout
+        end
+      end
+
```
Comment (rhenium) [inline lib/openssl/ssl.rb:435]: SocketForwarder (in this file, above) is the right place for this definition.

**Comment by ioquatix** [comment]
> Also, a single blocking operation on TLS, such as #connect and #sysread, can involve multiple blocking operations on the underlying socket. This would mean we can't rely on RUBY_IO_TIMEOUT_DEFAULT, but have to get rb_io_timeout() and subtract elapsed time before waiting again.

Thanks for this clarification. `IO#timeout` was designed with discrete, low level operations in mind. I think we can start with the simplest implementation, which is discrete timeout per operation. The goal of `IO#timeout` is essentially a safety net. If `#connect` performs several internal operations, it's reasonable that there are several "timeouts". Other parts of `IO` behave similarly. e.g. each call to `wait_whatever` will use a new timeout, therefore, on slow connections, it's possible for things like `gets` to invoke read several times, and thus wait several times, each with a new timeout.

If users desire application level timeouts, e.g. "This group of operations should take at most X seconds", they should use `Timeout.timeout` or some similar construct.

**Comment by rhenium** [comment]
This is superseded by #714 which fixes SSLSocket to correctly raise `IO::Timeout`.

**Pull request closed without merging**